File: app/src/main/java/de/tuberlin/mcc/simra/app/activities/OpenBikeSensorActivity.java
package de.tuberlin.mcc.simra.app.activities;

import android.app.AlertDialog;
import android.content.BroadcastReceiver;
import android.content.pm.PackageManager;
import android.content.res.ColorStateList;
import android.graphics.Color;
import android.os.Bundle;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.Button;
import android.widget.LinearLayout;
import android.widget.RadioButton;
import android.widget.RadioGroup;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;

import java.util.HashSet;
import java.util.Set;

import de.tuberlin.mcc.simra.app.R;
import de.tuberlin.mcc.simra.app.databinding.ActivityOpenbikesensorBinding;
import de.tuberlin.mcc.simra.app.services.OBSService;
import de.tuberlin.mcc.simra.app.util.PermissionHelper;
import de.tuberlin.mcc.simra.app.util.SharedPref;
import pl.droidsonroids.gif.GifImageView;


public class OpenBikeSensorActivity extends AppCompatActivity {
    private static final String TAG = "OpenBikeSensorActivity_LOG";
    BroadcastReceiver receiver;
    Set<BluetoothDevice> foundDevices;
    BluetoothDevice selectedDevice;
    ActivityOpenbikesensorBinding binding;
    RadioGroup devices;
    private AlertDialog alertDialog;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        binding = ActivityOpenbikesensorBinding.inflate(LayoutInflater.from(this));
        setContentView(binding.getRoot());

        foundDevices = new HashSet<>();
        initializeToolBar();
        Log.d(TAG, "onCreate");
        binding.retryButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                startScanningDevices();
                foundDevices = new HashSet<>();
            }
        });
        Button connectButton = new Button(this);
        connectButton.setVisibility(View.GONE);
        connectButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                if (selectedDevice != null) {
                    OBSService.connectDevice(OpenBikeSensorActivity.this, selectedDevice.deviceId);
                    Log.d(TAG, "connecting to device: " + selectedDevice.deviceId);
                }
            }
        });
        devices = new RadioGroup(this);
        devices.setOrientation(RadioGroup.VERTICAL);
        devices.setOnCheckedChangeListener(new RadioGroup.OnCheckedChangeListener() {
            @Override
            public void onCheckedChanged(RadioGroup radioGroup, int id) {
                for (BluetoothDevice device : foundDevices) {
                    if (device.hashCode() == id) {
                        selectedDevice = device;
                        connectButton.setText("Connect to " + selectedDevice.deviceName);
                        connectButton.setVisibility(View.VISIBLE);
                    }
                }

            }
        });
        binding.handleBarWidth.setMaxValue(60);
        binding.handleBarWidth.setMinValue(0);
        binding.handleBarWidth.setValue(SharedPref.Settings.Ride.OvertakeWidth.getHandlebarWidth(this));
        binding.handleBarWidth.setOnValueChangedListener((numberPicker, oldVal, newVal) -> {
            SharedPref.Settings.Ride.OvertakeWidth.setTotalWidthThroughHandlebarWidth(newVal, this);
        });

        binding.btnDisconnect.setOnClickListener(view -> OBSService.disconnectAndUnpairDevice(this));
        OBSService.ConnectionState currentState = OBSService.getConnectionState();
        updateUI(currentState);
        if (!currentState.equals(OBSService.ConnectionState.CONNECTED)) {
            startScanningDevices();
        }

        binding.connectDevicesLayout.addView(devices);
        binding.connectDevicesLayout.addView(connectButton);
    }

    private void setClosePassBarColor(int distanceInCm) {
        int maxColorValue = Math.min(distanceInCm, 200); // 200 cm ist maximum, das grün
        // Algoritmus found https://stackoverflow.com/questions/340209/generate-colors-between-red-and-green-for-a-power-meter
        // Da n zwischen 0 -100 liegen soll und das maximum 200 ist, dann halbieren immer den Wert.
        int normalizedValue = maxColorValue / 2;
        int red = (255 * (100 - normalizedValue)) / 100;
        int green = (255 * normalizedValue) / 100;
        int blue = 0;
        // Color und Progress sind abhängig
        binding.progressBarClosePass.setProgressTintList(ColorStateList.valueOf(Color.rgb(red, green, blue)));
        binding.progressBarClosePass.setProgress(normalizedValue);
    }

    private void createRadioButton(BluetoothDevice device) {
        RadioButton radioButton = new RadioButton(this);
        radioButton.setText(device.deviceName);
        radioButton.setId(device.hashCode());
        devices.addView(radioButton);
    }

    private void showRetryButton() {
        binding.searchingCircle.setVisibility(View.GONE);
        binding.retryButton.setVisibility(View.VISIBLE);
    }

    private void hideRetryButton() {
        binding.searchingCircle.setVisibility(View.VISIBLE);
        binding.retryButton.setVisibility(View.GONE);
    }

    private void updateUI(OBSService.ConnectionState state) {
        switch (state) {
            case PAIRING:
                binding.deviceLayout.setVisibility(View.GONE);
                binding.connectDevicesLayout.setVisibility(View.GONE);
                binding.pairingLayout.setVisibility(View.VISIBLE);
                showTutorialDialog();
                break;
            case CONNECTED:
                binding.deviceLayout.setVisibility(View.VISIBLE);
                binding.connectDevicesLayout.setVisibility(View.GONE);
                binding.pairingLayout.setVisibility(View.GONE);
                closeTutorialDialog();
                break;
            case CONNECTION_REFUSED:
                showRetryButton();
            case DISCONNECTED:
                showRetryButton();
            case SEARCHING:
                binding.deviceLayout.setVisibility(View.GONE);
                binding.connectDevicesLayout.setVisibility(View.VISIBLE);
                binding.pairingLayout.setVisibility(View.GONE);
                hideRetryButton();
            default:
                break;
        }
    }

    private void registerReceiver() {
        receiver = OBSService.registerCallbacks(this, new OBSService.OBSServiceCallbacks() {
            @Override
            public void onDeviceFound(String deviceName, String deviceId) {
                BluetoothDevice foundDevice = new BluetoothDevice(deviceName, deviceId);
                if (!foundDevices.contains(foundDevice)) {
                    foundDevices.add(foundDevice);
                    createRadioButton(foundDevice);

                }
            }

            @Override
            public void onConnectionStateChanged(OBSService.ConnectionState newState) {
                Log.d(TAG, "connection state: " + newState.name());
                updateUI(newState);
            }

            @Override
            public void onDistanceValue(OBSService.Measurement value) {
                Log.d(TAG, "value: " + value.toString());
                int distance = -1;
                if (value != null && value.leftSensorValues.size() > 0) {
                    distance = value.leftSensorValues.get(0);
                    binding.deviceInfoTextView.setText(getString(R.string.obs_activity_text_last_distance) + " " + distance + " cm");
                    setClosePassBarColor(distance);
                }
            }
        });

    }

    private void startScanningDevices() {
        binding.devicesList.removeAllViews();
        OBSService.startScanning(this);
    }

    private void initializeToolBar() {
        setSupportActionBar(binding.toolbar.toolbar);
        getSupportActionBar().setDisplayShowTitleEnabled(false);
        binding.toolbar.toolbar.setTitle("");
        binding.toolbar.toolbar.setSubtitle("");
        binding.toolbar.toolbarTitle.setText(getString(R.string.obs));

        binding.toolbar.backButton.setOnClickListener(v -> finish());
    }

    private void connectToDevice(String deviceId) {
        OBSService.connectDevice(this, deviceId);
    }

    private void showTutorialDialog() {
        AlertDialog.Builder alert = new AlertDialog.Builder(this);
        alert.setTitle(getString(R.string.obsConnecting));
        alert.setMessage(getString(R.string.obsTutorial));

        LinearLayout gifLayout = new LinearLayout(this);
        LinearLayout.LayoutParams gifMargins = new LinearLayout.LayoutParams(LinearLayout.LayoutParams.WRAP_CONTENT, LinearLayout.LayoutParams.WRAP_CONTENT);
        GifImageView gif = new GifImageView(this);
        gif.setImageResource(R.drawable.tutorial);
        gif.setVisibility(View.VISIBLE);
        gifMargins.setMargins(50, 0, 50, 0);
        gif.setLayoutParams(gifMargins);
        gifLayout.addView(gif);
        alert.setView(gifLayout);
        alert.setPositiveButton("Ok", (dialog, whichButton) -> {
        });
        alertDialog = alert.show();

    }

    private void closeTutorialDialog() {
        if (alertDialog != null)
            alertDialog.dismiss();
    }

    @Override
    protected void onPause() {
        OBSService.unRegisterCallbacks(receiver, this);
        super.onPause();
    }

    @Override
    protected void onResume() {
        registerReceiver();
        OBSService.ConnectionState currentState = OBSService.getConnectionState();
        if (!currentState.equals(OBSService.ConnectionState.CONNECTED)) {
            startScanningDevices();
        }
        super.onResume();
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
    }

    private class BluetoothDevice {
        private String deviceName;
        private String deviceId;

        public BluetoothDevice(String deviceName, String deviceId) {
            this.deviceName = deviceName;
            this.deviceId = deviceId;
        }

        @Override
        public int hashCode() {
            return deviceName.hashCode() * deviceId.hashCode();
        }

        @Override
        public boolean equals(Object obj) {
            if (this == obj)
                return true;
            if (obj == null)
                return false;
            if (getClass() != obj.getClass())
                return false;
            BluetoothDevice other = (BluetoothDevice) obj;
            if (!deviceName.equals(other.deviceName))
                return false;
            if (!deviceId.equals(other.deviceId))
                return false;
            return true;
        }
    }
}

File: app/src/main/java/de/tuberlin/mcc/simra/app/activities/OpenBikeSensorActivity.kt
package de.tuberlin.mcc.simra.app.activities

import android.Manifest
import android.bluetooth.BluetoothAdapter
import android.bluetooth.BluetoothManager
import android.content.Context
import android.content.Intent
import android.content.pm.PackageManager
import android.content.res.ColorStateList
import android.graphics.Color
import android.os.Build
import android.os.Bundle
import android.util.Log
import androidx.activity.result.ActivityResultLauncher
import androidx.core.app.ActivityCompat
import de.tuberlin.mcc.simra.app.R
import de.tuberlin.mcc.simra.app.databinding.ActivityOpenbikesensorBinding
import de.tuberlin.mcc.simra.app.util.BaseActivity
import de.tuberlin.mcc.simra.app.util.ConnectionManager
import de.tuberlin.mcc.simra.app.util.ConnectionManager.BLESTATE
import de.tuberlin.mcc.simra.app.util.ConnectionManager.CLOSE_PASS_CHARACTERISTIC_UUID
import de.tuberlin.mcc.simra.app.util.ConnectionManager.SENSOR_DISTANCE_CHARACTERISTIC_UUID
import de.tuberlin.mcc.simra.app.util.PermissionHelper.REQUEST_ENABLE_BT
import de.tuberlin.mcc.simra.app.util.PermissionHelper.hasBLEPermissions
import de.tuberlin.mcc.simra.app.util.PermissionHelper.requestBlePermissions
import de.tuberlin.mcc.simra.app.util.SharedPref
import de.tuberlin.mcc.simra.app.util.Utils
import de.tuberlin.mcc.simra.app.util.Utils.activityResultLauncher
import de.tuberlin.mcc.simra.app.util.ble.ConnectionEventListener


private const val TAG = "OpenBikeSensorActivityCM_LOG"
private const val ENABLE_BLUETOOTH_REQUEST_CODE = 1

class OpenBikeSensorActivity : BaseActivity() {
    private lateinit var binding: ActivityOpenbikesensorBinding


    private val notifyingCharacteristicsToSubscribeTo = listOf(SENSOR_DISTANCE_CHARACTERISTIC_UUID, CLOSE_PASS_CHARACTERISTIC_UUID)
    private var blestate = BLESTATE.DISCONNECTED
    private var deviceName = "---"
    private lateinit var activityResultLauncher:ActivityResultLauncher<Intent>


    private fun updateUI() {
        runOnUiThread {
            Log.d(TAG, "updateUI() - blestate: $blestate")
            when (blestate) {
                BLESTATE.DISCONNECTED -> {
                    binding.bluetoothButton.text = getString(R.string.obs_activity_button_start_scan)
                    binding.statusText.text = getString(R.string.obs_activity_text_start)
                }
                BLESTATE.FOUND -> {
                    binding.bluetoothButton.text = getString(R.string.obs_activity_button_connect_device)
                    binding.statusText.text = getString(R.string.obs_activity_text_connect,deviceName)
                }
                BLESTATE.CONNECTED -> {
                    binding.bluetoothButton.text = getString(R.string.obs_activity_button_disconnect_device)
                    binding.statusText.text = getString(R.string.obs_activity_text_disconnect,deviceName)
                }
                else -> {
                    binding.bluetoothButton.text = getString(R.string.obs_activity_button_stop_scan)
                    binding.statusText.text = getString(R.string.obs_activity_text_stop)
                }
            }
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        activityResultLauncher = activityResultLauncher(this@OpenBikeSensorActivity)
        binding = ActivityOpenbikesensorBinding.inflate(layoutInflater)
        val view = binding.root
        setContentView(view)
        initializeToolBar()
        if (ConnectionManager.bleState == BLESTATE.CONNECTED) {
            blestate = BLESTATE.CONNECTED
            deviceName = ConnectionManager.scanResult.device.name
            updateUI()
        }

        binding.bluetoothButton.setOnClickListener {

            Log.d(TAG, "pressed button. blestate: $blestate")
            when (blestate) {
                BLESTATE.DISCONNECTED -> {
                    if (!hasBLEPermissions(this)) {
                        requestBlePermissions(this@OpenBikeSensorActivity, REQUEST_ENABLE_BT)
                    } else {
                        ConnectionManager.startScan(this)
                    }
                }
                BLESTATE.FOUND -> ConnectionManager.connect(notifyingCharacteristicsToSubscribeTo,this)
                BLESTATE.CONNECTED -> ConnectionManager.disconnect(ConnectionManager.scanResult.device)
                else -> ConnectionManager.stopScan()
            }
        }

        // handlebar width
        binding.handleBarWidth.maxValue = 60
        binding.handleBarWidth.minValue = 0
        binding.handleBarWidth.value = SharedPref.Settings.Ride.OvertakeWidth.getHandlebarWidth(this)
        binding.handleBarWidth.setOnValueChangedListener { picker, oldVal, newVal ->
            SharedPref.Settings.Ride.OvertakeWidth.setTotalWidthThroughHandlebarWidth(newVal, this)
        }
    }

    private val bluetoothAdapter: BluetoothAdapter by lazy {
        val bluetoothManager = getSystemService(Context.BLUETOOTH_SERVICE) as BluetoothManager
        bluetoothManager.adapter
    }

    override fun onResume() {
        super.onResume()

        if (!bluetoothAdapter.isEnabled) {
            promptEnableBluetooth()
        }
        ConnectionManager.registerListener(connectionEventListener)
    }

    override fun onPause() {
        super.onPause()
        ConnectionManager.unregisterListener(connectionEventListener)
    }

    private fun promptEnableBluetooth() {
        if (!bluetoothAdapter.isEnabled) {
            val enableBtIntent = Intent(BluetoothAdapter.ACTION_REQUEST_ENABLE)
            activityResultLauncher.launch(enableBtIntent)
        }
    }

    private val connectionEventListener by lazy {
        ConnectionEventListener().apply {
            onScanStart = {
                Log.d(TAG, "connectionEventListener: onScanStart")
                blestate = BLESTATE.SEARCHING
                updateUI()
            }
            onDeviceFound = {
                Log.d(TAG, "connectionEventListener: onDeviceFound")
                deviceName = it.name
                blestate = BLESTATE.FOUND
                updateUI()
            }
            onScanStop = {
                Log.d(TAG, "connectionEventListener: onScanStop")
                if (!it) {
                    blestate = BLESTATE.DISCONNECTED
                    updateUI()
                }
            }
            onConnectionSetupComplete = {
                Log.d(TAG, "connectionEventListener: onConnectionSetupComplete")
                deviceName = it.device.name
                SharedPref.App.OpenBikeSensor.setObsDeviceName(deviceName, this@OpenBikeSensorActivity)
                blestate = BLESTATE.CONNECTED
                updateUI()
            }
            onDisconnect = {
                Log.d(TAG, "connectionEventListener: onDisconnect")
                deviceName = "---"
                SharedPref.App.OpenBikeSensor.deleteObsDeviceName(this@OpenBikeSensorActivity)
                blestate = BLESTATE.DISCONNECTED
                updateUI()
            }
            onSensorDistanceNotification = {
                binding.deviceInfoTextView.text = this@OpenBikeSensorActivity.getString(R.string.obs_activity_text_last_distance,it.leftDistance)
                setColePassBarColor(it.leftDistance.toInt())
            }
            onClosePassNotification = {
                Log.d(TAG, "ClosePass - time: ${it.obsTime} left distance: ${it.leftDistance} right distance: ${it.rightDistance}")
            }
            onTimeRead = {
                Log.d(TAG, "Time - time: $it")
            }
        }
    }

    private fun setColePassBarColor(distance: Int) {
        val maxColorValue = distance.coerceAtMost(200) // maximum: 200cm (green)
        val normalizedValue = maxColorValue / 2
        val red = (255 * (100 - normalizedValue)) / 100
        val green = (255 * normalizedValue) / 100
        val blue = 0
        // Color and Progress are dependant of distance
        binding.progressBarClosePass.progressTintList = ColorStateList.valueOf(Color.rgb(red,green,blue))
        binding.progressBarClosePass.progress = normalizedValue
    }

    private fun initializeToolBar() {
        setSupportActionBar(binding.toolbar.toolbar)
        supportActionBar!!.setDisplayShowTitleEnabled(false)
        binding.toolbar.toolbar.title = ""
        binding.toolbar.toolbar.subtitle = ""
        binding.toolbar.toolbarTitle.text = getString(R.string.obs)
        binding.toolbar.backButton.setOnClickListener { v -> finish() }
    }

}